app/grandchallenge/algorithms/models.py

```diff
@@ -521,6 +558,28 @@ def form_field_label(self):
         return title
 
 
+class AlgorithmInterfaceThroughTable(models.Model):
+    algorithm = models.ForeignKey(Algorithm, on_delete=models.CASCADE)
+    interface = models.ForeignKey(AlgorithmInterface, on_delete=models.CASCADE)
+    is_default = models.BooleanField(default=False)
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["algorithm"],
+                condition=Q(is_default=True),
+                name="unique_default_interface_per_algorithm",
+            ),
+            models.UniqueConstraint(
+                fields=["algorithm", "interface"],
```

Comment:
Is this necessary? I thought that `ManyToMany` would handle this?

Reply:
Yes, this is necessary, or else the through table can contain multiple rows for the same interface and algorithm. Why would a ManyToMany handle this be default?

Reply:
Because methods like `algorithm.inputs.add(foo)` are idempotent.

Reply:
Ah ok, I added the constraint based on the documentation (see the first info box [here](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.ManyToManyField.through)), but maybe I misunderstood that comment? Testing it out in the shell, it looks like the add and remove methods are still idempotent even when using a custom through table.... 

In any case, though, the idempotency only holds for the `add` and `remove` methods. You can still create duplicates by directly using the through table to create them (if you don't have a constraint in place) `AlgorithmInterfaceThroughTable.objects.create(interface=..., algorithm-alg)`. So I would say it's safer to add the constraint, right?

Reply:
That also leads me to a secondary question: is it then preferred to add interfaces to algorithm using the `add` method (with a `through_defaults` specified to populate the `is_default` field), rather than doing what I do now in the data migration: `AlgorithmInterfaceThroughTable.objects.create(algorithm=algorithm, interface=io, is_default=True)`?

Reply:
> So I would say it's safer to add the constraint, right?

Yes definitely, I didn't notice that box. We should add them for all existing through tables.

> Is it then preferred to add interfaces to algorithm using the add method

I would say so, for normal m2m fields you do not have access to the through table, so here you are making the assumption that the class is the correct through table (which is pretty safe but you never know!). I think it is best to be consistent. Generally, we shouldn't have to access through tables directly.

Reply:
One more question: is the naming of the new through tables consistent with the existing ones in the codebase?

Reply:
> One more question: is the naming of the new through tables consistent with the existing ones in the codebase?

No, the "ThroughTable" part should go if we want to be consistent with the rest of the codebase. I had added that bit to clearly distinguish it from the `AlgorithmInterface` model (and to avoid a name clash). We could call it `InterfaceAlgorithm` (which implies that it's something else to me). What would you suggest for the name? 

The other two that I added today are more straightforward: `AlgorithmInterfaceInput` and `AlgorithmInterfaceOutput`

Reply:
`AlgorithmInputInterface` (with `algorithm` and `input_interface` defined) and `AlgorithmOutputInterface` (with `algorithm` and `output_interface` defined) wouldn't clash and would be consistent, right?

Or, maybe this is too confusing but we could pre-empt our renaming to `AlgorithmInputSlot` (with `algorithm` and `input_slot` defined) and `AlgorithmOutputSlot` (with `algorithm` and `output_slot` defined)?

Reply:
I think I wasn't clear enough earlier. There are 3 through tables that need names. 

Normally, we follow `Model1Model2` naming convention for through tables. So then for the through table with `algorithm_interface` and `algorithm` defined, we'd end up with either `AlgorithmInterfaceAlgorithm` or `AlgorithmAlgorithmInterface` - both sound terrible to me. Pre-empting our naming change would not affect the name because the `AlgorithmInterface` model would still be called that after the naming change. 

For the other 2 (linking `algorithm_interface` and `input` / `output` respectively), we could already adopt the slot terminology, but it might be a bit confusing if the naming migration is not taken up in the near future... The names I suggested above `AlgorithmInterfaceInput` and `AlgorithmInterfaceOutput` do not clash with anything and would be an option for the time being.

Reply:
Okay then it's probably fine.